<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <!--[if lt IE 9]>
    <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->
  <style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
  </style>
</head>
<body>
<h1 id="homework-4">Homework 4</h1>
<p>Task: You have been given <a href="http://hyphaltip.github.io/GEN220_2015/data/Scer_Trinity.fasta.gz">a set of RNA sequences</a> that are assembled. You would like to investigate how many of them have poly-A tails. Look for sequences that have AAAAAA or AATAAA, find the 3' most of these (e.g. the last one in the sequence) <a href="http://nar.oxfordjournals.org/content/27/24/4751.full">this paper</a> talks about finding motifs related to poly-A positioning.</p>
<p>Here is a file of assembled mRNA transcripts from an RNAseq experiment in FASTA format.</p>
<ul>
<li>Write a script to read in the data, and count which sequences have polyA sites -- you can use <a href="this%20script">http://hyphaltip.github.io/GEN220_2015/Examples/read_fasta.py</a> as a starting point for reading the data</li>
<li>Generate a distribution of polyA lengths (distance between the motif you found and the end of the contig)</li>
<li>Compute summary statistics for these lengths (mean, median)</li>
<li><p>Plot histogram of this distribution - using R. Here is a <a href="http://hyphaltip.github.io/GEN220_2015/Examples/histogram.R">simple R script</a> to plot histogram you can run like this - just make sure you file is called 'polyA_lengths.dat' or change the code in the R script.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">R</span> --no-save <span class="kw">&lt;</span> histogram.R</code></pre></div></li>
</ul>
<p>You may want to revisit the regular expression lectures and see this page on <a href="https://docs.python.org/2/library/re.html">Regular Expressions from Python</a>.</p>
<ol start="2" type="1">
<li>A restriction enzyme cuts DNA at specific locations. Identify the number of cut sites in the genome of Bacillus subtilis of the EcoRI (GAATTC) motif. Your program should simply print out</li>
</ol>
<pre class="text"><code>Genome file: [MY FILE]
Number of EcoRI (GAATTC) cut sites: [Number of sites]
...</code></pre>
<p>Here is a list of RE sites - you will need to re-write some of these to convert to a regular expression.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">EcoRI   <span class="op">=</span> <span class="st">&quot;GAATTC&quot;</span> 
Bsu15I  <span class="op">=</span> <span class="st">&quot;ATCGAT&quot;</span>  
Bsu36I  <span class="op">=</span> <span class="st">&quot;CCTNAGG&quot;</span>
BsuRI   <span class="op">=</span> <span class="st">&quot;GGCC&quot;</span>
EcoRII  <span class="op">=</span> <span class="st">&quot;CCWGG&quot;</span></code></pre></div>
<p>The abbreviation for DNA ambiguity patterns can be found on <a href="https://en.wikipedia.org/wiki/Nucleic_acid_notation">this page</a>.</p>
<p>Use the <em>B. subtilis</em> genome <a href="http://www.ncbi.nlm.nih.gov/genome/genomes/665">here</a> and you can find a direct link to the strain 168 <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000009045.1_ASM904v1">FTP folder of the assembly</a>. You want the &quot;.fna&quot; file.</p>
<p>You can also use the E. coli K-12 genome from NCBI or what is <a href="http://hyphaltip.github.io/GEN220_2015/data/Ecoli_K-12.fasta">available here</a>.</p>
<p>The goal of this is to write generic code so you are welcome to run this on any genome really. There are many sequenced strains, it would be interesting to compare if the number of cut sites (or their size) varied among strains.</p>
<p>For the advanced programmers. Think about (or try) to make your program handle a folder of sequence files to read and provide a report.</p>
<ol start="3" type="1">
<li>Write a program to find which proteins have a Nuclear Localization Signal in yeast. The <a href="http://www.jbc.org/content/282/27/19292.full">following paper Fries et al</a>. Figure 8 demonstrates a sequence pattern. There is one part that is ambiguous (X)n where n varies some in the species. Feel free to write this targeted for the <em>Saccharomyces cerevisiae</em> search or just make it some arbitrarily long length based on all the data. Here is the <a href="http://www.yeastgenome.org/download-data/sequence">sequence download site</a> for Saccharomyces and <a href="http://downloads.yeastgenome.org/sequence/S288C_reference/orf_protein/orf_trans_all.fasta.gz">here are the proteins in fasta format</a> you can download.</li>
</ol>
<ul>
<li>You can spot check how you tool does by looking at the Gene Ontology (GO) <a href="http://downloads.yeastgenome.org/curation/literature/go_slim_mapping.tab">data for Saccharomyces</a>. The 4th column indicates the ontology - a C there means 'cellular localization' so the rows that have a 'C' in the 4th column give you some indication of where the protein localizes to (if it is known). Typically a nuclear localization signal would indicate the protein localizes to the nucleus. So you could see how many of your hits have this prediction. Or you could test how many DON'T have this prediction since this signal sequence may not have been used to classify localization in GO</li>
</ul>
</body>
</html>
