<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<h1 id="variant-analysis-projects">Variant analysis Project(s)</h1>
<p>Chemical mutagenesis of a plant and mutants selected for those with the mutation. Here is the publication on the dataset <a href="http://journal.frontiersin.org/article/10.3389/fpls.2013.00362/abstract">Facile mutant identification via a single parental backcross method and application of whole genome sequencing based mapping pipelines</a> Develop a strategy to start with raw sequence data, identify the mutations, and report the number of changes in genes or in intergenic regions. Can you identify the mutational pattern of EMS in the DNA?</p>
<p>Given another set of individuals sequenced, identify the differences in naturally occuring mutations.</p>
<h1 id="transcriptome-analysis-projects">Transcriptome analysis Project(s)</h1>
<p>Build a library of full length transcripts from a particular organism (TBD) from raw RNAseq.</p>
<p>Identify the number of genes which are novel (unannotated) in the current genome annotation.</p>
<p>Generate graphs of the average size or location of these genes.</p>
<h1 id="phylogenetics">Phylogenetics</h1>
</body>
</html>
